src/hidpp20.h

```diff
@@ -127,6 +127,126 @@ int hidpp20_batterylevel_get_battery_level(struct hidpp20_device *device,
 					   uint16_t *next_level);
 
 /* -------------------------------------------------------------------------- */
+/* 0x1300: LED software control                                               */
+/* -------------------------------------------------------------------------- */
+
+#define HIDPP_PAGE_LED_SW_CONTROL              0x1300
+
+enum hidpp20_led_sw_ctrl_led_type {
+	HIDPP20_LED_TYPE_BATTERY = 0x01,
+	HIDPP20_LED_TYPE_DPI,
+	HIDPP20_LED_TYPE_PROFILE,
+	HIDPP20_LED_TYPE_LOGO,
+	HIDPP20_LED_TYPE_COSMETIC
+};
+
+enum hidpp20_led_sw_ctrl_led_mode {
+	HIDPP20_LED_MODE_OFF = 0x01,
+	HIDPP20_LED_MODE_ON = 0x02,
+	HIDPP20_LED_MODE_BLINK = 0x04,
+	HIDPP20_LED_MODE_TRAVEL = 0x08,
+	HIDPP20_LED_MODE_RAMP_UP = 0x10,
+	HIDPP20_LED_MODE_RAMP_DOWN = 0x20,
+	HIDPP20_LED_MODE_HEARTBEAT = 0x40,
+	HIDPP20_LED_MODE_BREATHING = 0x80
+};
+
+#define HIDPP20_LED_SW_CONTROL_LED_INDEX_ALL 0xff
+
+struct hidpp20_led_sw_ctrl_led_info {
+	uint8_t index;
+	uint16_t type;
+	uint8_t physical_count;
+	uint16_t caps;
+	uint8_t nvconfig_caps;
+} __attribute__((packed));
+_Static_assert(sizeof(struct hidpp20_led_sw_ctrl_led_info) == 7, "Invalid size");
+
+struct hidpp20_led_sw_ctrl_led_state {
+	uint8_t index;
+	uint16_t mode;
+	union {
+		struct {
+			uint16_t max_brightness;
+			uint16_t period;
+			uint16_t timeout;
+		} breathing;
+		struct {
+			uint16_t unused;
+			uint16_t delay;
+		} traveling;
+		struct {
+			uint16_t index;
+			uint16_t on_time;
+			uint16_t off_time;
+		} blink;
+		struct {
+			// Which physical LED to turn on
```

Comment:
The G502 spectrum DPI LED has 3 physical LEDs, but the on.index accepted values are in the 1–5 range (for the 5 DPI index). The profile LED also has 3 physical LEDs (actually the same 3, they are shared), and the accepted value range is 1–3 (for 3 profiles). I don't think this index controls the physical LEDs directly.

Reply:
It may not be direct but setting this determines which physical LED(s) is/are lit up. Though I agree the formulation is not the clearest ;)